### PR TITLE
Salto 5973 create a fix element that will remove transition rules dependent on missing jira apps

### PR DESCRIPTION
**Pull request by IlayDavid95**
Jira:

- Added a 'fix element' that will **_remove_** workflow transition rules of missing (not installed in target environment) Jira apps. This 'fix element' is under the feature flag of config.deploy.ignoreMissingExtensions in Jira NaCl.

---

_Additional context for reviewer_

Part of a series of PRs. This one depends on #5943, so review that one first.

---
_Release Notes_: 

Jira:

- Added a feature flag under config.deploy called ignoreMissingExtensions (defaults to false). When enabled, workflow transition rules of Jira apps that aren't installed in the target environment will be removed.

### CODE REVIEW

File: packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    const fixedTransitions: ElemID[] = []
+
+    const fixedElements = await awu(elements)
+      .filter(isInstanceElement)
+      .map(instance => instance.clone())
+      .filter(isWorkflowV2Instance)
+      .filter(workflowInstance =>
+        _.some(
+          Object.entries(workflowInstance.value.transitions).map(([transitionName, transition]) => {
+            if (fixedWorkflowTransition(installedExtensionsMap, transition)) {
```
Comment (shayc331) [inline packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts:111]: it is not clear that a function that called `fixedWorkflowTransition` is changing the instance and returns a boolean.
I think we should split the transition fix, first getting the new transitions without changing the instance, and then understand which transition were changed and inform the user accordingly
lets talk f2f

**Review by shayc331 — COMMENTED** [review]
looking great, had some comments

File: packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts (hunk 1 above)
Comment (IlayDavid95) [reply packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts]: changed to updateAndCheck like we discussed

File: packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+  // Fix validator rules
+  const oldValidators = transition.validators
+  if (oldValidators !== undefined) {
+    transition.validators = filterOrEmpty(isTransitionRuleNotFromMissingExtension, transition.validators)
+    wasChanged = wasChanged || transition.validators.length !== oldValidators.length
+  }
+
+  // Fix condition rules
+  const fixedConditions = fixedWorkflowConditionGroup(isTransitionRuleNotFromMissingExtension, transition.conditions)
```
[the remark this answers is not in the document]
Comment (IlayDavid95) [reply packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts]: change method name to have prefix `updateAndCheck` like we discussed.

File: packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+  const isTransitionRuleNotFromMissingExtension = (transitionRule: WorkflowV2TransitionRule): boolean => {
+    const extensionId = getExtensionIdFromWorkflowTransitionRule(transitionRule)
+    return extensionId === undefined ? true : extensionId in installedExtensionsMap
+  }
+
+  let wasChanged = false
+
+  // Fix action rules
+  const oldActions = transition.actions
+  if (oldActions !== undefined) {
```
[the remark this answers is not in the document]
Comment (IlayDavid95) [reply packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts]: Extracted to some funky typescript function :)

File: packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+  WorkflowV2TransitionConditionGroup,
+  WorkflowV2Transition,
+  WorkflowV2TransitionRule,
+  isWorkflowV2Instance,
+} from '../filters/workflowV2/types'
+import { ExtensionType } from '../client/client'
+
+const { awu } = collections.asynciterable
+
+const filterOrEmpty = <T>(predicate: (value: T) => boolean, list?: T[]): T[] =>
```
[the remark this answers is not in the document]
Comment (IlayDavid95) [reply packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts]: removed method.

File: packages/jira-adapter/test/fix_elements/remove_missing_extension_transition_rules.test.ts [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+
+    const result = await removeMissingExtensionsTransitionRulesHandler({ config, client, elementsSource })([instance])
+    expect(result).toEqual({ fixedElements: [], errors: [] })
+  })
+  it('should fix transitions with action rules of missing extensions', async () => {
+    instance.value.transitions.transition1 = createSkeletonWorkflowV2Transition('transition1')
+    instance.value.transitions.transition1.actions = [...NON_EXISTENT_EXTENSIONS_TRANSITION_RULES]
+
+    const result = await removeMissingExtensionsTransitionRulesHandler({ config, client, elementsSource })([instance])
+    expect(result.fixedElements).toHaveLength(1)
```
Comment (shayc331) [inline packages/jira-adapter/test/fix_elements/remove_missing_extension_transition_rules.test.ts:128]: please check the content of the fixed transition, here and in the rest of the tests

File: packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+  installedExtensionsMap: Record<string, ExtensionType>,
+  transition: WorkflowV2Transition,
+): boolean => {
+  const checkExtensionForRule = (transitionRule: WorkflowV2TransitionRule): boolean => {
+    const extensionId = getExtensionIdFromWorkflowTransitionRule(transitionRule)
+    return extensionId === undefined ? true : extensionId in installedExtensionsMap
+  }
+
+  // Fix action rules
+  const fixedActions = updateAndCheckTransitionRules(transition, 'actions', checkExtensionForRule)
```
Comment (shayc331) [inline packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts:76]: this name does not indicate it is a boolean, please change it to something like `isFixedActions` or similar, same for the rest

File: packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+const updateAndCheckWorkflowConditionGroup = (
+  predicate: (transitionRule: WorkflowV2TransitionRule) => boolean,
+  conditionGroup?: WorkflowV2TransitionConditionGroup,
+): boolean => {
+  if (conditionGroup === undefined) {
+    return false
+  }
+  const fixedConditions = updateAndCheckTransitionRules(conditionGroup, 'conditions', predicate)
```
Comment (shayc331) [inline packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts:55]: please use a name that indicates it is a boolean

File: packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+  if (conditionGroup.conditionGroups !== undefined) {
+    const fixedConditionGroups = conditionGroup.conditionGroups.map(innerConditionGroup =>
```
Comment (shayc331) [inline packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts:58]: same naming issue here

File: packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+  WorkflowV2Transition,
+  WorkflowV2TransitionRule,
+  isWorkflowV2Instance,
+} from '../filters/workflowV2/types'
+import { HasKeyOfType, MakeOptional } from '../../../lowerdash/src/types'
+import { getExtensionIdFromWorkflowTransitionRule } from '../common/workflow/transition_rules'
+
+const { awu } = collections.asynciterable
+
+const updateAndCheckTransitionRules = <
```
Comment (shayc331) [inline packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts:34]: please add a comment that explain the behavior here

File: packages/lowerdash/src/types.ts [hunk 10]
```diff
@@ -35,10 +35,19 @@ export const hasMember = <T, M extends keyof T>(m: M, o: T): o is HasMember<T, M
 export const filterHasMember = <T, M extends keyof T>(m: M, objs: T[]): HasMember<T, M>[] =>
   objs.filter(f => hasMember(m, f)) as HasMember<T, M>[]
 
+// Ensures an object contains a specific key K of type T, allowing for additional properties.
+export type HasKeyOfType<K extends string, T> = { [P in K]: T } & Record<string, unknown>
```
Comment (shayc331) [inline packages/lowerdash/src/types.ts:39]: @dantal4 can you please take a look on those functions?

**Review by shayc331 — COMMENTED** [review]
looks very good, had some small comments

File: packages/lowerdash/src/types.ts (hunk 10 above)
Comment (dantal4) [inline packages/lowerdash/src/types.ts:39]: If I'm not wrong additional properties are always allowed- no need to specify the `Record<string, unknown>`. The only problem is to **directy** initiate a variable of type `HasKeyOfType<K extends string, T>` with fields that aren't in `K`:

```typescript
type HasKeyOfType<K extends string, T> = { [P in K]: T }

const t1: HasKeyOfType<'abc'|'def', number> = {
  abc: 1,
  def: 2,
  q: 4, // doesn't work :(
}

const t2 = {
  abc: 1,
  def: 2,
  q: 4,
}

const t3: HasKeyOfType<'abc'|'def', number> = t // works!

```

File: packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+            return false
+          }),
+        ),
+      )
+      .filter(values.isDefined)
+      .toArray()
+
+    const errors = fixedTransitions.map(id => ({
+      elemID: id,
+      severity: 'Info' as const,
```
[the remark this answers is not in the document]
Comment (IlayDavid95) [reply packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts:117]: While I agree, the linter sadly does not :(

File: packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+      message: 'Deploying workflow transition without all of its rules.',
+      detailedMessage:
+        'This workflow transition contains rules for Jira apps that do not exist in the target environment. It will be deployed without them.',
+    }))
```
[the remark this answers is not in the document]
Comment (IlayDavid95) [reply packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts:121]: We discussed this with @liorn and decided to remain on the transition level.

File: packages/lowerdash/src/types.ts [hunk 13]
```diff
@@ -35,10 +35,19 @@ export const hasMember = <T, M extends keyof T>(m: M, o: T): o is HasMember<T, M
...
+// Extracts keys from type T whose values are assignable to type U.
 export type KeysOfType<T, U> = { [K in keyof T]: T[K] extends U ? K : never }[keyof T]
+// Extracts keys from type T whose values extend type U.
 export type KeysOfExtendingType<T, U> = { [K in keyof T]: U extends T[K] ? K : never }[keyof T]
+// Converts the keys of type T into an enum-like type where each key is a required property.
 export type TypeKeysEnum<T> = Required<{ [k in keyof T]: k }>
 
+// Utility type to make a specific key K optional in type T
+export type MakeOptional<T, K extends keyof T> = Omit<T, K> & Partial<Pick<T, K>>
```
Comment (dantal4) [inline packages/lowerdash/src/types.ts:48]: I'm not sure that this type is necessary. see my next comment 👇

File: packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+const { awu } = collections.asynciterable
+
+const updateAndCheckTransitionRules = <
+  K extends string,
+  T extends MakeOptional<HasKeyOfType<K, WorkflowV2TransitionRule[]>, K>,
+>(
+  obj: T,
+  key: K,
+  predicate: (value: WorkflowV2TransitionRule) => boolean,
+): boolean => {
```
Comment (dantal4) [inline packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts:41]: does this work? it's clearer IMO

```suggestion
  K extends string,
>(
  obj: Partial<HasKeyOfType<K, WorkflowV2TransitionRule[]>>,
  key: K,
  predicate: (value: WorkflowV2TransitionRule) => boolean,
): boolean => {
```

**Review by dantal4 — COMMENTED** [review]
I added a few comments about lowerdash types 👍

File: packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+/**
+ * This fixer makes sure that there are no workflow transitions with rules dependent on Jira apps not installed in the target environment.
+ */
+export const removeMissingExtensionsTransitionRulesHandler: FixElementsHandler =
+  ({ client, config }) =>
+  async (elements: Element[]) => {
+    if (!config.fetch.enableNewWorkflowAPI || !config.deploy.ignoreMissingExtensions) {
+      return { fixedElements: [], errors: [] }
+    }
+    const installedExtensionsMap: Record<string, ExtensionType> = await getInstalledExtensionsMap(client)
```
[the remark this answers is not in the document]
Comment (IlayDavid95) [reply packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts:94]: done

File: packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+          Object.entries(workflowInstance.value.transitions).map(([transitionName, transition]) => {
+            if (updateAndCheckWorkflowTransition(installedExtensionsMap, transition)) {
+              fixedTransitions.push(workflowInstance.elemID.createNestedID('transitions', transitionName))
+              return true
+            }
+            return false
+          }),
+        ),
+      )
+      .filter(values.isDefined)
```
[the remark this answers is not in the document]
Comment (IlayDavid95) [reply packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts:112]: No, you're right.

File: packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts (hunk 9 above)
Comment (IlayDavid95) [reply packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts:34]: Done.

File: packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts (hunk 8 above)
Comment (IlayDavid95) [reply packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts:58]: Done.

File: packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts (hunk 14 above)
Comment (IlayDavid95) [reply packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts:41]: Doesn't partial make all properties optional? I wanted to make only one optional.

Comment (dantal4) [reply packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts:41]: `HasKeyOfType` or this function doesn't know about the other fields of the input object, so it doesn't matter

File: packages/lowerdash/src/types.ts (hunk 10 above)
Comment (IlayDavid95) [reply packages/lowerdash/src/types.ts:39]: Removed the Record<string, unknown>

File: packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts (hunk 14 above)
Comment (IlayDavid95) [reply packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts:41]: I guess you're right in this context, removed the type.

File: packages/lowerdash/src/types.ts (hunk 13 above)
Comment (IlayDavid95) [reply packages/lowerdash/src/types.ts:48]: Removed.

File: packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts (hunk 6 above)
Comment (IlayDavid95) [reply packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts:76]: Done.

File: packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts (hunk 7 above)
Comment (IlayDavid95) [reply packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts:55]: Done.

File: packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts (hunk 6 above)
Comment (shayc331) [reply packages/jira-adapter/src/fix_elements/remove_missing_extension_transition_rules.ts:85]: please change the names here as well

File: packages/jira-adapter/test/fix_elements/remove_missing_extension_transition_rules.test.ts [hunk 17]
```diff
@@ -0,0 +1,289 @@
...
+    expect(result).toEqual({ fixedElements: [], errors: [] })
+  })
+  it('should fix transitions with action rules of missing extensions', async () => {
+    instance.value.transitions.transition1 = createSkeletonWorkflowV2Transition('transition1')
+    instance.value.transitions.transition1.actions = [...NON_EXISTENT_EXTENSIONS_TRANSITION_RULES]
+
+    const result = await removeMissingExtensionsTransitionRulesHandler({ config, client, elementsSource })([instance])
+
+    const fixedInstance = instance.clone()
+    fixedInstance.value.transitions.transition1.actions = []
```
Comment (shayc331) [inline packages/jira-adapter/test/fix_elements/remove_missing_extension_transition_rules.test.ts:130]: please add a test (or modify the exist tests) when there are more than one transitionRule, and after the fixElements the list is not empty

**Review by shayc331 — APPROVED** [review]
looking great 🏅 
had some small comments approving to save a cycle
please wait for @dantal4 approval

**Comment by dantal4** [comment]
> looking great 🏅 had some small comments approving to save a cycle please wait for @dantal4 approval

lowerdash usage LGTM 👍

File: packages/jira-adapter/test/fix_elements/remove_missing_extension_transition_rules.test.ts (hunk 17 above)
Comment (IlayDavid95) [reply packages/jira-adapter/test/fix_elements/remove_missing_extension_transition_rules.test.ts:130]: done.

File: packages/jira-adapter/test/fix_elements/remove_missing_extension_transition_rules.test.ts (hunk 5 above)
Comment (IlayDavid95) [reply packages/jira-adapter/test/fix_elements/remove_missing_extension_transition_rules.test.ts]: Done.
